src/services/AudioMixer.cxx

```diff
@@ -0,0 +1,208 @@
+*/
+
+void AudioMixer::setMusicVolume(VolumeLevel volume) noexcept
+{
+  GetService<GameLoopMQ>().push(AudioMusicVolumeChangeEvent{volume}); 
+}
+
+void AudioMixer::setSoundEffectVolume(VolumeLevel volume) noexcept 
+{ 
+  GetService<GameLoopMQ>().push(AudioSoundVolumeChangeEvent{volume}); 
+}
+
+void AudioMixer::play(SoundtrackID&& ID) noexcept { GetService<GameLoopMQ>().push(AudioPlayEvent{ID}); }
+
+void AudioMixer::play(AudioTrigger&& trigger) noexcept 
+{ 
+  GetService<GameLoopMQ>().push(AudioTriggerEvent{trigger}); 
+}
+
+void AudioMixer::play(SoundtrackID&& ID, Coordinate3D&& position) noexcept 
+{ 
+  GetService<GameLoopMQ>().push(AudioPlay3DEvent{ID, position});
+}
+
+void AudioMixer::play(AudioTrigger&& trigger, Coordinate3D&& position) noexcept
+{
+  GetService<GameLoopMQ>().push(AudioTrigger3DEvent{trigger, position}); 
+}
+
+void AudioMixer::setMuted(bool isMuted) noexcept
+{
+  GetService<GameLoopMQ>().push(AudioSetMutedEvent{isMuted}); 
+}
+
+/*
++------------------+
+|  Event Handlers  |
++------------------+
+*/
+
+void AudioMixer::handleEvent(const AudioTriggerEvent&& event)
+{
+  auto& possibilities = m_Triggers[event.trigger];
+  if(possibilities.size() == 0) 
+  {
+    LOG() << "Warning: no Soundtracks are triggered by " << event.trigger._to_string();
+    return;
+  }
+  SoundtrackID& trackID = *GetService<Randomizer>().choose(possibilities.begin(), possibilities.end());
+  SoundtrackUPtr& track = *m_GetSound.at(trackID);
+  playSoundtrack(track);
+}
+
+/* @todo Implement this */
+void AudioMixer::handleEvent(const AudioTrigger3DEvent&& event)
+{
+  string ErrorMsg = "Unimplemented Error: ";
+  ErrorMsg += __PRETTY_FUNCTION__;
+  throw RuntimeError(ErrorMsg);
+}
+
+/* @todo Implement this */
+void AudioMixer::handleEvent(const AudioPlayEvent&& event)
+{
+  SoundtrackUPtr& track = *m_GetSound.at(event.ID);
+  playSoundtrack(track);
+}
+
+/* @todo Implement this */
+void AudioMixer::handleEvent(const AudioPlay3DEvent&& event)
+{
+  string ErrorMsg = "Unimplemented Error: ";
+  ErrorMsg += __PRETTY_FUNCTION__;
+  throw RuntimeError(ErrorMsg);
+}
+
+/* @todo Implement this */
+void AudioMixer::handleEvent(const AudioSoundVolumeChangeEvent&& event)
+{
+  string ErrorMsg = "Unimplemented Error: ";
+  ErrorMsg += __PRETTY_FUNCTION__;
+  throw RuntimeError(ErrorMsg);
+}
+
+/* @todo Implement this */
+void AudioMixer::handleEvent(const AudioMusicVolumeChangeEvent&& event)
+{
+  string ErrorMsg = "Unimplemented Error: ";
+  ErrorMsg += __PRETTY_FUNCTION__;
+  throw RuntimeError(ErrorMsg);
+}
+
+/* @todo Implement this */
+void AudioMixer::handleEvent(const AudioSetMutedEvent&& event)
+{
+  string ErrorMsg = "Unimplemented Error: ";
+  ErrorMsg += __PRETTY_FUNCTION__;
+  throw RuntimeError(ErrorMsg);
+}
+
+/*
++-----------+
+|  Helpers  |
++-----------+
+*/
+
+void AudioMixer::playSoundtrack(SoundtrackUPtr& track)
+{
+  if(!track)
+    throw RuntimeError("Received an invalid soundtrack");
+  if(track->Channel.get() == -1)
```

Comment:
This assumption doesn't hold if the track was played before (it would keep its previous `ChannelID`)

- [ ] Add a `Mix_ChannelFinished` callback that will set `track->Channel` back to `-1`